smithy-aws-endpoints/src/main/java/software/amazon/smithy/rulesengine/aws/traits/StandardRegionalEndpointsTrait.java

```diff
@@ -0,0 +1,200 @@
+        regionSpecialCases = builder.regionSpecialCases.copy();
+    }
+
+    /**
+     * Gets the map of partition string to a list of partition special cases defined in the trait.
+     *
+     * @return Returns a map of partition string to a list of {@link PartitionSpecialCase}
+     */
+    public Map<String, List<PartitionSpecialCase>> getPartitionSpecialCases() {
+        return partitionSpecialCases;
+    }
+
+    /**
+     * Gets the map of partition string to a list of region special cases defined in the trait.
+     *
+     * @return Returns a map of region string to a list of {@link RegionSpecialCase}
+     */
+    public Map<String, List<RegionSpecialCase>> getRegionSpecialCases() {
+        return regionSpecialCases;
+    }
+
+    @Override
+    protected Node createNode() {
+        ObjectNode.Builder partitionSpecialCasesNodeBuilder = ObjectNode.objectNodeBuilder();
+        for (Map.Entry<String, List<PartitionSpecialCase>> entry: partitionSpecialCases.entrySet()) {
+            List<Node> nodes = new ArrayList<>();
+            for (PartitionSpecialCase partitionSpecialCase : entry.getValue()) {
+                nodes.add(partitionSpecialCase.toNode());
+            }
+            partitionSpecialCasesNodeBuilder.withMember(entry.getKey(), Node.fromNodes(nodes));
+        }
+
+        ObjectNode.Builder regionSpecialCasesNodeBuilder = ObjectNode.objectNodeBuilder();
+        for (Map.Entry<String, List<RegionSpecialCase>> entry: regionSpecialCases.entrySet()) {
+            List<Node> nodes = new ArrayList<>();
+            for (RegionSpecialCase regionSpecialCase : entry.getValue()) {
+                nodes.add(regionSpecialCase.toNode());
+            }
+            regionSpecialCasesNodeBuilder.withMember(entry.getKey(), Node.fromNodes(nodes));
+        }
+
+        return ObjectNode.objectNodeBuilder()
+                .sourceLocation(getSourceLocation())
+                .withMember(PARTITION_SPECIAL_CASES, partitionSpecialCasesNodeBuilder.build())
+                .withMember(REGION_SPECIAL_CASES, partitionSpecialCasesNodeBuilder.build())
+                .build();
+    }
+
+    @Override
+    public SmithyBuilder<StandardRegionalEndpointsTrait> toBuilder() {
+        return new Builder()
+                .partitionSpecialCases(partitionSpecialCases)
+                .regionSpecialCases(regionSpecialCases);
+    }
+
+    public static Builder builder() {
+        return new Builder();
+    }
+
+    public static final class Provider extends AbstractTrait.Provider {
+        public Provider() {
+            super(ID);
+        }
+
+        @Override
+        public Trait createTrait(ShapeId target, Node value) {
+            ObjectNode objectNode = value.expectObjectNode();
+
+            StandardRegionalEndpointsTrait.Builder builder = builder()
+                    .sourceLocation(value);
+
+            Optional<ObjectNode> memberNode = objectNode.getObjectMember(PARTITION_SPECIAL_CASES);
```

Comment:
Should use `objectNode.containsMember` in the if statement, since this `Optional` value is discarded.